scripts/perfcontrib.py

```diff
@@ -0,0 +1,115 @@
+from json import loads, dumps
+from urllib.request import urlopen, Request
+from urllib.parse import urlencode
+from urllib.error import HTTPError
+import time
+import sys
+import os
+
+tenantId = "72f988bf-86f1-41af-91ab-2d7cd011db47"
+appId = "c2fe4cd0-be4a-468b-aa4f-078c67dcab6e"
+
+uploadService = "https://perfcontrib.azurewebsites.net"
+uploadEndpoint = f"{uploadService}/api/UploadPerfData"
+authEndpoint = f"{uploadService}/.auth/login/aad"
+authDetailsEndpoint = f"{uploadService}/.auth/me"
+
+aadUrl = f"https://login.microsoftonline.com/{tenantId}"
+
+def get_token() -> str:
+    path = os.path.expanduser("~/.perfcontrib")
+
+    token: str | None = None
+
+    try:
+        if not os.path.exists(path):
+            os.makedirs(path)
+        with open(os.path.expanduser("~/.perfcontrib/token")) as tokenfile:
+            token = tokenfile.readline()
+    except FileNotFoundError:
+        pass
+
+    if token:
+        try:
+            with urlopen(Request(authDetailsEndpoint,
+                    headers = { "X-ZUMO-AUTH": token })) as response:
+                print("Using cached credentials.")
+        except HTTPError as error:
+            token = None
+
+    if not token:
+        token = authenticate()
+    
+    if token:
+        with open(os.path.expanduser("~/.perfcontrib/token"), "w") as tokenfile:
+            tokenfile.write(token)
+    
+    return token
+
+def authenticate() -> str:
+    authBody = {
+        "tenant": tenantId,
+        "client_id": appId,
+        "scope": "User.Read openid profile",
+    }
+    
+    authBodyEncoded = urlencode(authBody).encode()
+
+    with urlopen(Request(f"{aadUrl}/oauth2/v2.0/devicecode", data = authBodyEncoded)) as response:
+        devicecodeResponse = loads(response.read().decode('utf-8'))
+
+    print(devicecodeResponse["message"])
+    
+    authBody2 = {
+        "tenant": tenantId, 
+        "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
+        "client_id": appId,
+        "device_code": devicecodeResponse["device_code"]
+    }
+    
+    authBody2Encoded = urlencode(authBody2).encode()
+    
+    authStatus = "waiting"
+    while (authStatus == "waiting"):
+        # Try to get the access token. if we encounter an error check the reason. 
```

Comment:
I don't see any code here that checks the reason if there is an error, but it does make sense to be added if it does provide useful error information on failures.